akka-http-core/src/test/scala/akka/http/scaladsl/RenegotiatingHttpsTestServer.scala

```diff
@@ -0,0 +1,75 @@
+/**
+ * Copyright (C) 2009-2014 Typesafe Inc. <http://www.typesafe.com>
+ */
+
+package akka.http.scaladsl
+
+import java.security.Principal
+
+import akka.http.impl.util.ExampleHttpContexts
+import akka.http.scaladsl.model.headers.RequestClientCertificate
+
+import scala.concurrent.duration._
+import scala.concurrent.Await
+import akka.actor.ActorSystem
+import akka.http.scaladsl.model._
+import akka.stream.ActorMaterializer
+import com.typesafe.config.{ ConfigFactory, Config }
+import HttpMethods._
+
+object RenegotiatingHttpsTestServer extends App {
+  val testConf: Config = ConfigFactory.parseString("""
+    akka.loglevel = INFO
+    akka.log-dead-letters = off
+    akka.http.server.parsing.tls-session-info-header = on
+    """)
+  implicit val system = ActorSystem("ServerTest", testConf)
+  implicit val fm = ActorMaterializer()
+
+  try {
+    val binding = Http().bindAndHandleSync({
+      case HttpRequest(GET, Uri.Path("/"), _, _, _) ⇒ index
+      case req @ HttpRequest(GET, Uri.Path("/reneg"), _, _, _) ⇒
+        def responseForPrincipal(principal: Principal): HttpResponse =
+          HttpResponse(entity = s"Hello ${principal.getName}!")
+
+        req.header[headers.`Tls-Session-Info`] match {
+          case Some(infoHeader) if infoHeader.info.peerPrincipal.isDefined ⇒
+            responseForPrincipal(infoHeader.info.peerPrincipal.get)
+          case _ ⇒
+            HttpResponse(headers = RequestClientCertificate { info ⇒
+              info.peerPrincipal match {
+                case Some(p) ⇒ responseForPrincipal(p)
+                case None    ⇒ HttpResponse(StatusCodes.Unauthorized, entity = "401 Requested peer certificate but didn't get any")
+              }
```

Comment:
I’m not sure about this continuation: why not just use the same route again with new info next time? (or maybe I’m misunderstanding)